### PR TITLE
feat: HTML to markdown parser

### PR DESCRIPTION
**Pull request by pranshuchittora**
@roryabraham @jasperhuangg @marcaaron  will you please review this?

Context: https://github.com/Expensify/Expensify.cash/issues/2847

# Tests
**What unit/integration tests cover your change? What autoQA tests cover your change?**
Yes I have added tests

**What tests did you perform that validates your changed worked?**
Yes

# QA
**What does QA need to do to validate your changes?**
Try various HTML strings with `<br>` tags in various combinations

HTML -> MD
1. `Hello<br/>There` -> `Hello\nThere`
2. `Hello<br></br>There` -> `Hello\nThere`

Tests -> https://github.com/Expensify/expensify-common/pull/381/files#diff-7031bf36a0d060dfeb1742c45b8fa6853fb5790ef4da366b0fb96ceaafd342ea

**What areas to they need to test for regressions?**
Parser

### CODE REVIEW

**Comment by pranshuchittora** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by roryabraham** [comment]
@pranshuchittora I've run the test check suite on this PR and found that there are lint errors that need to be resolved

**Comment by pranshuchittora** [comment]
@roryabraham I have fixed the linting issues :)

File: lib/ExpensiMark.js [hunk 1]
```diff
@@ -157,6 +157,23 @@ export default class ExpensiMark {
...
+        this.htmlToMarkdownRules = [
+            {
+                name: 'newline',
+
+                // Replaces open and closing <br><br/> tags with a single <br/>
+                pre: inputString => inputString.replace('<br><br/>', '<br/>'),
+                regex: /<br\s*[\\/]?>/gi,
+                replacement: '\n'
+            },
+
```
Comment (Luke9389) [inline lib/ExpensiMark.js:175]: Please remove this extra line of whitespace.

File: lib/ExpensiMark.js [hunk 2]
```diff
@@ -157,6 +157,23 @@ export default class ExpensiMark {
...
+         *
+         * @type {Object[]}
+         */
+        this.htmlToMarkdownRules = [
+            {
+                name: 'newline',
+
+                // Replaces open and closing <br><br/> tags with a single <br/>
+                pre: inputString => inputString.replace('<br><br/>', '<br/>'),
+                regex: /<br\s*[\\/]?>/gi,
```
Comment (Luke9389) [inline lib/ExpensiMark.js:172]: I have a few questions regarding our regex pattern: 
 - Why are we attempting to detect `<br\>`? 
 - What is the purpose of using the case insensitive flag?

File: lib/ExpensiMark.js [hunk 3]
```diff
@@ -250,4 +267,23 @@ export default class ExpensiMark {
 
         return replacedText;
     }
+
+    /**
+     * Replaces html with elements
```
Comment (Luke9389) [inline lib/ExpensiMark.js:272]: Can we change this to `Replaces HTML with markdown`?

File: lib/ExpensiMark.js [hunk 4]
```diff
@@ -250,4 +267,23 @@ export default class ExpensiMark {
...
+     *
+     * @param {String} htmlString
+     *
+     * @returns {String}
+     */
+    htmlToMarkdown(htmlString) {
+        let generatedMarkdown = htmlString;
+        this.htmlToMarkdownRules.forEach((rule) => {
+            // Pre-processes input HTML before applying regex
```
Comment (Luke9389) [inline lib/ExpensiMark.js:281]: Please add an empty line before this comment (like the one above).

**Review by Luke9389 — CHANGES_REQUESTED** [review]
Hello @pranshuchittora. Nice work so far. I've got a few small comments and one larger one that I'll post below.

**Comment by Luke9389** [comment]
Do we want to match breaking space tags that are inside of code blocks? 

If you sent a message like the following, what would we want our code to do? 
"Hey fellow engineer, I noticed we are missing a `<br/>` closing tag on line X." 

Currently, I think we'd end up matching and replacing that tag, which is not desired behavior. Perhaps we should establish a method for preventing matches inside of code blocks. Thoughts?

**Comment by Luke9389** [comment]
As a side note, I think you can put the changes from this PR here as well. I like the organization of having them separate, but I think it's best for our workflow to just have one PR to track.

File: lib/ExpensiMark.js (hunk 2 above)
Comment (pranshuchittora) [reply lib/ExpensiMark.js:172]: Hi, 
1. we can remove `\` regex
2. This is because HTML is a case insensitive language 

I am thinking of adding `<br></br>` as well in `pre`

**Comment by pranshuchittora** [comment]
> "Hey fellow engineer, I noticed we are missing a <br/> closing tag on line X."

As this happens on runtime (conversion), therefore this becomes irrelevant. 
My approach / philosophy is inspired by the way HTML works instead of breaking it assumes and fixes the tree on runtime

File: lib/ExpensiMark.js (hunk 2 above)
Comment (Luke9389) [reply lib/ExpensiMark.js:172]: Ok great! Thx. This should work nicely, idk if the `pre` is necessary.

Comment (pranshuchittora) [reply lib/ExpensiMark.js:172]: Initially having a regex worked, but later when I was writing some tests and figuring out edge cases. It became necessary to add pre.

Usually, some sanitisers work by appending a closing tag for the uncloses tags `<br>`. That is the reason behind adding pre. Or you can say my addiction to defensive programming 😅

Comment (pranshuchittora) [reply lib/ExpensiMark.js:172]: I am also accounting for both HTML style and react style of closing

File: lib/ExpensiMark.js (hunk 4 above)
Comment (pranshuchittora) [reply lib/ExpensiMark.js:281]: Screenshot from 2021-06-05 20-53-10

Comment (Luke9389) [reply lib/ExpensiMark.js:280]: Ah, yea, that's due to the whitespace on that line. If you backspace a bunch of times it'll get rid of that linting error.

**Comment by Luke9389** [comment]
> As this happens on runtime (conversion), therefore this becomes irrelevant. My approach / philosophy is inspired by the way HTML works instead of breaking it assumes and fixes the tree on runtime

Interesting. Would you mind elaborating a bit further, for my sake? Also, is this something we can test?

**Comment by pranshuchittora** [comment]
> > As this happens on runtime (conversion), therefore this becomes irrelevant. My approach / philosophy is inspired by the way HTML works instead of breaking it assumes and fixes the tree on runtime
> 
> Interesting. Would you mind elaborating a bit further, for my sake? Also, is this something we can test?

Sure, what I understand is that we should show some warning for unclosed `<br>` in dev mode. This kinda becomes as messages are going to be generated by the users on runtime. 

The only way this becomes relevant is when input to HTML conversion happens. Instead of handling it here, I recommend writing robust test cases for the HTML conversion parser.

**Comment by Luke9389** [comment]
> Sure, what I understand is that we should show some warning for unclosed `<br>` in dev mode.
 
Ok it's possible we're talking about two different things right now. My main concern is that if someone puts a br tag inside backticks, that we won't make a newline. I'm not sure what you're referring to with showing a warning.

**Comment by pranshuchittora** [comment]
> > Sure, what I understand is that we should show some warning for unclosed `<br>` in dev mode.
> 
> Ok it's possible we're talking about two different things right now. My main concern is that if someone puts a br tag inside backticks, that we won't make a newline. I'm not sure what you're referring to with showing a warning.

Yeah, I agree that's an edge case. I try finding out how other parsers are doing this unfortunately one of the famous parser has buggy behaviour with this as well.

http://domchristie.github.io/turndown/

Paste this HTML 👇🏼 
```
<code>
Hello There <br/>
</code>
```

**Review by jasperhuangg — APPROVED** [review]

**Comment by Luke9389** [comment]
Couldn't we just use regex to accomplish this? I think negative lookahead will allow us to avoid matching a breaking tag wrapped in backticks. Something like this: <pre><br\s*[/]?>(?!`)</pre>

That way we won't match `<br/>`

**Comment by pranshuchittora** [comment]
> I think negative lookahead will allow us to avoid matching a breaking tag wrapped in backticks

@Luke9389 backticks are from the markdown world.
Backticks are converted to `<code>` in HTML. The input to this will look like this
```html
<code>
Hello There <br/>
</code>

```

File: lib/ExpensiMark.js (hunk 4 above)
Comment (Luke9389) [reply lib/ExpensiMark.js:280]: Did this get resolved?

**Comment by pranshuchittora** [comment]
Hi, @marcaaron can you help merging this

**Comment by marcaaron** [comment]
Sorry, @pranshuchittora please give the assigned reviewers time to handle this and if > 3 days ask in the Slack channel for help

**Comment by Luke9389** [comment]
Hey @pranshuchittora, sorry for the late response here. I've been on vacation the past week. 

Ok, I see what you're saying about the backticks, but we aren't out of the weeds yet.

First, I want to be sure we're on the same page. When someone sends a message with this, `<br/>`, they're talking about a br tag, and so we shouldn't parse that to a new line. That gets sent, turned into HTML (< code >< br/ >< /code >), and now when we are parsing that HTML back to markdown, we want to be sure we don't create a new line (we want it to be turned into `<br/>`). 

So in my view, we should have some way of preventing the content inside of code blocks from being parsed. Maybe that should be saved for a different PR? What do you think @pranshuchittora?

**Review by jasperhuangg — CHANGES_REQUESTED** [review]
I agree with Luke's comments about ensuring that any code inside of code blocks isn't parsed.

**Comment by pranshuchittora** [comment]
> Hey @pranshuchittora, sorry for the late response here. I've been on vacation the past week.
> 
> Ok, I see what you're saying about the backticks, but we aren't out of the weeds yet.
> 
> First, I want to be sure we're on the same page. When someone sends a message with this, `<br/>`, they're talking about a br tag, and so we shouldn't parse that to a new line. That gets sent, turned into HTML (< code >< br/ >< /code >), and now when we are parsing that HTML back to markdown, we want to be sure we don't create a new line (we want it to be turned into `<br/>`).
> 
> So in my view, we should have some way of preventing the content inside of code blocks from being parsed. Maybe that should be saved for a different PR? What do you think @pranshuchittora?

Yup we are on the same page 👍🏼 
Solving this is very tricky as we can’t find `<code>` directly as it can have some attribute like `<code some-attribute="xyz" >`. I checked few parsers out there as discussed in this comment https://github.com/Expensify/expensify-common/pull/381#issuecomment-856948421

IMO let’s solve that with another issue and PR

**Comment by Luke9389** [comment]
Yea, I think solving that in another PR is reasonable. It'll be its own issue with its own solution. I'm ready to merge as soon as this gets resolved: https://github.com/Expensify/expensify-common/pull/381/files#r644268511

File: lib/ExpensiMark.js (hunk 4 above)
Comment (pranshuchittora) [reply lib/ExpensiMark.js:280]: I tried but it's not working 

Screenshot from 2021-06-15 22-53-26

**Comment by pranshuchittora** [comment]
> Yea, I think solving that in another PR is reasonable. It'll be its own issue with its own solution. I'm ready to merge as soon as this gets resolved: https://github.com/Expensify/expensify-common/pull/381/files#r644268511

https://github.com/Expensify/expensify-common/pull/381/files#r652000032

File: lib/ExpensiMark.js (hunk 4 above)
Comment (Luke9389) [reply lib/ExpensiMark.js:280]: Hmm, OK. Not a big deal I guess. Thanks for trying

**Review by Luke9389 — APPROVED** [review]
One more thing, can you add specific tests (numbered and detailed) for what Web QA should do to test this? Try and make it as easy for them as possible to understand what they need to do. Right now it's too vague.

**Review by jasperhuangg — APPROVED** [review]
